linkml/generators/pydanticgen/array.py

```diff
@@ -0,0 +1,396 @@
+        """Array shape specified with ``dimensions`` without additional anonymous axes"""
+        pass
+
+    @abstractmethod
+    def mixed_shape(self, array: ArrayExpression) -> SlotResult:
+        """Array shape with both ``dimensions`` and a ``max_number_dimensions`` for anonymous axes"""
+        pass
+
+
+class ListOfListsArray(ArrayRangeGenerator):
+    """
+    Represent arrays as lists of lists!
+
+    TODO: Move all validation of values (eg. anywhere we raise a ValueError) to the ArrayExpression
+    dataclass and out of the generator class
+    """
+
+    REPR = ArrayRepresentation.LIST
+
+    @staticmethod
+    def _list_of_lists(dimensions: int, dtype: str) -> str:
+        return ("List[" * dimensions) + dtype + ("]" * dimensions)
+
+    @staticmethod
+    def _labeled_dimension(dimension: DimensionExpression, dtype: str) -> SlotResult:
+        # TODO: Preserve label representation in some readable way! doing the MVP now of using conlist
+        if dimension.exact_cardinality and (dimension.minimum_cardinality or dimension.maximum_cardinality):
+            raise ValueError("Can only specify EITHER exact_cardinality OR minimum/maximum cardinality")
+        elif dimension.exact_cardinality:
+            dmin = dimension.exact_cardinality
+            dmax = dimension.exact_cardinality
+        elif dimension.minimum_cardinality or dimension.maximum_cardinality:
+            dmin = dimension.minimum_cardinality
+            dmax = dimension.maximum_cardinality
+        else:
+            # TODO: handle labels for labeled but unshaped arrays
+            return SlotResult(annotation="List[" + dtype + "]")
+
+        items = []
+        if int(PYDANTIC_VERSION[0]) >= 2:
+            if dmin is not None:
+                items.append(f"min_length={dmin}")
+            if dmax is not None:
+                items.append(f"max_length={dmax}")
+        else:
+            if dmin is not None:
+                items.append(f"min_items={dmin}")
+            if dmax is not None:
+                items.append(f"max_items={dmax}")
+        items.append(f"item_type={dtype}")
+        items = ", ".join(items)
+        annotation = f"conlist({items})"
+
+        return SlotResult(annotation=annotation, imports=_ConListImports)
+
+    def any_shape(self, array: Optional[ArrayExpression] = None, with_inner_union: bool = False) -> SlotResult:
+        """
+        An AnyShaped array
+
+        Args:
+            array (:class:`.ArrayExpression`): The array expression (not used)
+            with_inner_union (bool): If ``True`` , the innermost type is a ``Union`` of the ``AnyShapeArray`` class
+                and ``dtype`` (default: ``False`` )
+
+        """
+        if self.dtype in ("Any", "AnyType"):
+            annotation = "AnyShapeArray"
+        else:
+            annotation = f"AnyShapeArray[{self.dtype}]"
+
+        if with_inner_union:
+            annotation = f"Union[{annotation}, {self.dtype}]"
```

Comment:
I'm a bit confused here. This is used below to handle the case where `array.maximum_number_dimensions` is False. Let's say the dtype is integer. The resulting annotation is `Union[AnyShapeArray[integer], integer]` as opposed to just `AnyShapeArray[integer]`. What is the "inner union"? It seems like this line makes it so that scalar integers are allowed, but I am surely missing something.

Reply:
AnyShapeArray needs to be at least one level deep of a list when used on its own to validate. When specifying explicit dimensions, one needs to specify all those dimensions also as conlists or lists. This is to handle the inclusive part of the range - "three constrained dimensions or more" is 3 conlists, but if it was just an AnyShapeArray then it is minimum 4 dimensions. Adding the inner Union allows the "3" part of "3 or more"

Reply:
Alternatively we could allow AnyShapeArray to validate with a scalar value (0 dimensional array) and coerce it to a length 1 list. It just makes the validation behavior of AnyShapeArray less explicit, but I think that would also be a totally reasonable way to do it if we prefer that

Reply:
That makes sense. I understand now. Thanks!